File: examples/typescript/client_jwt_auth_flow.ts
/**
 * JWT auth flow — registration + login as pure stored functions.
 *
 * Demonstrates the canonical password-auth pattern using ekoDB stages:
 *
 *   register flow:  BcryptHash → Insert
 *   login flow:     FindOne  → BcryptVerify → If(ok) { JwtSign + Return }
 *   verify flow:    JwtVerify → If(claims != null) { 200 } else { 401 }
 *
 * Operator-owned secrets (the JWT signing key) flow through
 * `{{env.JWT_SECRET}}` so they never enter the function definition or
 * the LLM context window.
 *
 * Requires ekoDB >= 0.42.0.
 */

import { EkoDBClient, Stage, type UserFunction } from "@ekodb/ekodb-client";
import * as dotenv from "dotenv";

dotenv.config();

const BASE_URL = process.env.API_BASE_URL || "http://localhost:8080";
const API_KEY = process.env.API_BASE_KEY || "a-test-api-key-from-ekodb";

/** True when a save failed because the function label already exists (HTTP 409). */
function isAlreadyExistsError(error: unknown): boolean {
  const message = error instanceof Error ? error.message : String(error);
  return message.includes("status 409") || message.includes("already exists");
}

/** Idempotent save: create the function, or update it in place on 409. */
async function saveOrUpdateUserFunction(
  client: EkoDBClient,
  fn: UserFunction,
): Promise<void> {
  try {
    await client.saveUserFunction(fn);
  } catch (error) {
    if (!isAlreadyExistsError(error)) throw error;
    await client.updateUserFunction(fn.label, fn);
    console.log(`Function '${fn.label}' already existed — updated instead`);
  }
}

async function main() {
  const client = new EkoDBClient(BASE_URL, API_KEY);
  await client.init();
  console.log("✓ Client created");

  // 1. Register: bcrypt-hash, insert.
  const register: UserFunction = {
    label: "ts_users_register",
    name: "Register user",
    description: "Validate, bcrypt-hash, insert.",
    parameters: {
      email: { required: true },
      password: { required: true },
    },
    functions: [
      Stage.bcryptHash("{{password}}", "password_hash", 12),
      Stage.insert("ts_users", {
        email: "{{email}}",
        password_hash: "{{password_hash}}",
      }),
    ],
  };
  await saveOrUpdateUserFunction(client, register);
  console.log("✓ ts_users_register saved");

  // 2. Login: find user, verify bcrypt, sign JWT on success.
  const login: UserFunction = {
    label: "ts_users_login",
    name: "Login user",
    description: "Verify password, mint JWT.",
    parameters: {
      email: { required: true },
      password: { required: true },
    },
    functions: [
      Stage.findOne("ts_users", "email", "{{email}}"),
      Stage.bcryptVerify("{{password}}", "password_hash", "password_ok"),
      Stage.if(
        { type: "FieldEquals", value: { field: "password_ok", value: true } },
        [
          Stage.jwtSign(
            { email: "{{email}}" },
            "{{env.JWT_SECRET}}",
            "token",
            3600,
            "HS256",
          ),
          Stage.returnResponse({ ok: true, token: "{{token}}" }, 200),
        ],
        [Stage.returnResponse({ ok: false }, 401)],
      ),
    ],
  };
  await saveOrUpdateUserFunction(client, login);
  console.log("✓ ts_users_login saved");

  // 3. Verify a JWT — fail-closed when claims is null.
  const verify: UserFunction = {
    label: "ts_users_verify_token",
    name: "Verify JWT token",
    parameters: { token: { required: true } },
    functions: [
      // Synthetic record so JwtVerify has working_data[0] to read off.
      Stage.insert("_inflight_jwt_check", { token: "{{token}}" }, true, 60),
      Stage.jwtVerify("token", "{{env.JWT_SECRET}}", "claims", "HS256"),
      Stage.if(
        { type: "FieldEquals", value: { field: "claims", value: null } },
        [Stage.returnResponse({ ok: false }, 401)],
        [Stage.returnResponse({ ok: true }, 200)],
      ),
    ],
  };
  await saveOrUpdateUserFunction(client, verify);
  console.log("✓ ts_users_verify_token saved");

  console.log("\n=== Auth flow defined as pure stored functions ===");
  console.log("Call them like:");
  console.log(
    '  POST /api/functions/ts_users_register { "email": "a@b.com", "password": "s3cret" }',
  );
  console.log(
    '  POST /api/functions/ts_users_login    { "email": "a@b.com", "password": "s3cret" }',
  );
  console.log(
    '  POST /api/functions/ts_users_verify_token { "token": "<jwt>" }',
  );
  console.log(
    "\nSet JWT_SECRET in ekoDB's environment_vars whitelist before invoking.",
  );

  for (const label of [
    "ts_users_register",
    "ts_users_login",
    "ts_users_verify_token",
  ]) {
    await client.deleteUserFunction(label).catch(() => {});
  }
  console.log("\n✓ Cleaned up demo functions");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/python/client_jwt_auth_flow.py
#!/usr/bin/env python3
"""JWT auth flow — registration + login as pure stored functions.

Demonstrates the canonical password-auth pattern using ekoDB stages:

    register flow:  BcryptHash → Insert
    login flow:     FindOne  → BcryptVerify → If(ok) { JwtSign + Return }
    verify flow:    JwtVerify → If(claims != null) { 200 } else { 401 }

Operator-owned secrets (the JWT signing key) flow through
``{{env.JWT_SECRET}}`` so they never enter the function definition or
the LLM context window.

Requires ekoDB >= 0.43.0.
"""

import asyncio
import os
from pathlib import Path
from dotenv import load_dotenv

from ekodb_client import Client, Stage

env_path = Path(__file__).parent.parent / ".env"
load_dotenv(env_path)

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")


async def main() -> None:
    client = Client.new(BASE_URL, API_KEY)
    print("✓ Client created")

    # 1. Register: bcrypt-hash, insert.
    register = {
        "label": "py_users_register",
        "name": "Register user",
        "description": "Validate, bcrypt-hash, insert.",
        "parameters": {
            "email": {"required": True},
            "password": {"required": True},
        },
        "functions": [
            Stage.bcrypt_hash("{{password}}", "password_hash", cost=12),
            Stage.insert(
                "py_users",
                {"email": "{{email}}", "password_hash": "{{password_hash}}"},
            ),
        ],
    }
    try:
        await client.save_user_function(register)
        print("✓ py_users_register saved")
    except Exception as e:
        print(f"SaveUserFunction error: {e}")

    # 2. Login: find user, verify bcrypt, sign JWT on success.
    login = {
        "label": "py_users_login",
        "name": "Login user",
        "description": "Verify password, mint JWT.",
        "parameters": {
            "email": {"required": True},
            "password": {"required": True},
        },
        "functions": [
            {
                "type": "FindOne",
                "collection": "py_users",
                "key": "email",
                "value": "{{email}}",
            },
            Stage.bcrypt_verify("{{password}}", "password_hash", "password_ok"),
            Stage.if_condition(
                {
                    "type": "FieldEquals",
                    "value": {"field": "password_ok", "value": True},
                },
                then_functions=[
                    Stage.jwt_sign(
                        {"email": "{{email}}"},
                        "{{env.JWT_SECRET}}",
                        "token",
                        algorithm="HS256",
                        expires_in_secs=3600,
                    ),
                    Stage.return_response({"ok": True, "token": "{{token}}"}, 200),
                ],
                else_functions=[Stage.return_response({"ok": False}, 401)],
            ),
        ],
    }
    try:
        await client.save_user_function(login)
        print("✓ py_users_login saved")
    except Exception as e:
        print(f"SaveUserFunction error: {e}")

    # 3. Verify a JWT — fail-closed when claims is null.
    verify = {
        "label": "py_users_verify_token",
        "name": "Verify JWT token",
        "parameters": {"token": {"required": True}},
        "functions": [
            # Synthetic record so JwtVerify has working_data[0] to read off.
            Stage.insert(
                "_inflight_jwt_check",
                {"token": "{{token}}"},
                bypass_ripple=True,
                ttl=60,
            ),
            Stage.jwt_verify(
                "token", "{{env.JWT_SECRET}}", "claims", algorithm="HS256"
            ),
            Stage.if_condition(
                {"type": "FieldEquals", "value": {"field": "claims", "value": None}},
                then_functions=[Stage.return_response({"ok": False}, 401)],
                else_functions=[Stage.return_response({"ok": True}, 200)],
            ),
        ],
    }
    try:
        await client.save_user_function(verify)
        print("✓ py_users_verify_token saved")
    except Exception as e:
        print(f"SaveUserFunction error: {e}")

    print("\n=== Auth flow defined as pure stored functions ===")
    print("Call them like:")
    print(
        '  POST /api/functions/py_users_register { "email": "a@b.com", "password": "s3cret" }'
    )
    print(
        '  POST /api/functions/py_users_login    { "email": "a@b.com", "password": "s3cret" }'
    )
    print('  POST /api/functions/py_users_verify_token { "token": "<jwt>" }')

    for label in ("py_users_register", "py_users_login", "py_users_verify_token"):
        try:
            await client.delete_user_function(label)
        except Exception:
            pass
    print("\n✓ Cleaned up demo functions")


if __name__ == "__main__":
    asyncio.run(main())
